### PR TITLE
Add support for phi-3-vision-128k-instruct

### PR DESCRIPTION
**Pull request by JosefAlbers**
As per #28

### CODE REVIEW

**Comment by JosefAlbers** [comment]
Oh wow, thanks! Will do right away.

**Comment by Blaizzy** [comment]
Thanks for updating the test cases!

Looking at the model structure, I think we can sanitize the weights to create the following structure:

Model
- language_model
- vision_model -> vision_embed_tokens

This way we'll have a single structure:

```python
class Model(nn.Module):
    def __init__(self, config: ModelConfig):
        self.model_type = config.model_type
        self.config = config

        self.vision_model = VisionModel(config.vision_config)
        self.language_model = LanguageModel(config.text_config)


    def __call__(
        self,
        inputs: mx.array,
        pixel_values=None,
        image_sizes=None,
        cache=None,
    ):
       h = self.language_model.embed_tokens(inputs)
       p = np.argwhere(inputs < 0).tolist()
       if pixel_values is not None:
            h = self.vision_model(pixel_values, h, image_sizes, p)
       return self.language_model(h)

```

Example: https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/models/nanoLlava/nanoLlava.py#L240

**Comment by Blaizzy** [comment]
Hey @JosefAlbers 

Any updates?

**Comment by JosefAlbers** [comment]
@Blaizzy sorry, not yet.

**Comment by Blaizzy** [comment]
Hey @JosefAlbers 

Any updates?

**Comment by JosefAlbers** [comment]
Hi @Blaizzy, Thanks again for the feedback and for patiently guiding me through this process. I apologize for the delay in responding. 

After further reflection on the proposed model restructuring, however, I believe it might be more practical to defer this change for a future iteration. While I see the value in standardizing the model structure across the repository, the current structure aligns better with the original Phi-3-Vision model implementation, and maintaining this alignment could be beneficial in the short term.

My primary reason is to prioritize getting this model available for users as quickly as possible. There already are many newer and possibly better VLM models than Phi-3-Vision at this point (e.g., Kosmos-G, Florence, ...), and introducing a structural change now could potentially delay the merge, as it would require additional testing to ensure everything still functions as expected. I'd propose the following approach:

1. Merge the model in its current state.
2. Add a note to the issue tracker or project roadmap highlighting the potential for future refactoring to align with other models in the repository. 

This would allow us to revisit the structure at a later time when we have more bandwidth for potential adjustments and testing. I'm open to discussing this further and collaborating on the best path forward. Please let me know your thoughts on this approach.

Lastly, I want to thank you again for the great codes you've been writing in this repository. Your work has been an invaluable learning resource for me.

**Comment by Blaizzy** [comment]
Most welcome! 

Ayt, we can merge it and then update it.

Before that, can you rebase your branch and test it still works as expected.

**Comment by JosefAlbers** [comment]
@Blaizzy, Yeap, got a new branch rebased it up to date, copy pasted files from original to the new rebased branch, ran precommits and tests. I'm pretty exhausted at this point..🤞🏾

**Comment by Blaizzy** [comment]
Please rebase this branch with the main to fix the conflicts, run the tests and precommit.

`pytest .`

`pre-commit run --all-files`

**Comment by JosefAlbers** [comment]
Sorry for the silly error, I realize I might have not pushed after running the pre-commit on my local machine. I fixed them.

File: mlx_vlm/models/phi3_v/vision.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        txt_embeds=None,
+        img_sizes=None,
+        positions=None,
+        output_hidden_states=None,
+    ):
+        if output_hidden_states:
+            return self.img_processor.vision_model(
+                img_embeds, output_hidden_states=output_hidden_states
+            )
+        # print(0, txt_embeds.shape, img_embeds.shape, img_sizes.shape)
```
Comment (Blaizzy) [inline mlx_vlm/models/phi3_v/vision.py]:
```suggestion
        # print(0, txt_embeds.shape, img_embeds.shape, img_sizes.shape)
```
```suggestion
```

**Review by Blaizzy — CHANGES_REQUESTED** [review]
LGTM!

Just a tiny nit.

**Review by jjb21 — APPROVED** [review]
This line is likely from when I was struggling to understand and fix the model's broken outputs early in the porting process. I will delete it.

**Comment by Blaizzy** [comment]
Thank you very much @JosefAlbers for your hard work! 🎉👏🏾

We got some more models to port here: https://github.com/Blaizzy/mlx-vlm/issues/39

**Review by Blaizzy — APPROVED** [review]

**Comment by Blaizzy** [comment]
@JosefAlbers what's your twitter handle?

**Comment by JosefAlbers** [comment]
@Blaizzy You're very welcome, I'm just happy to help this awesome project grow. And I can't wait to see these new model ports in action!

**Comment by Blaizzy** [comment]
Thank you very much! 

You are an absolute legend 👏🏾

Please feel free to share more suggestions or pick any of the good first issues.

I'm currently working on the trainer 🚀

**Pull request merged**